File: android/src/main/java/me/carda/awesome_notifications/awesome_notifications_core/models/NotificationCalendarModel.java
package me.carda.awesome_notifications.awesome_notifications_core.models;

import android.content.Context;

import java.util.Calendar;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import me.carda.awesome_notifications.awesome_notifications_core.Definitions;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.AwesomeNotificationsException;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.ExceptionCode;
import me.carda.awesome_notifications.awesome_notifications_core.exceptions.ExceptionFactory;
import me.carda.awesome_notifications.awesome_notifications_core.utils.CronUtils;
import me.carda.awesome_notifications.awesome_notifications_core.utils.IntegerUtils;

public class NotificationCalendarModel extends NotificationScheduleModel {

    private static final String TAG = "NotificationCalendarModel";

    /// Field number for get and set indicating the era, e.g., AD or BC in the Julian calendar
    public Integer era;
    /// Field number for get and set indicating the year.
    public Integer year;
    /// Field number for get and set indicating the month.
    public Integer month;
    /// Field number for get and set indicating the day number within the current year (1-12).
    public Integer day;
    /// Field number for get and set indicating the hour of the day (0-23).
    public Integer hour;
    /// Field number for get and set indicating the minute within the hour (0-59).
    public Integer minute;
    /// Field number for get and set indicating the second within the minute (0-59).
    public Integer second;
    /// Field number for get and set indicating the millisecond within the second.
    public Integer millisecond;
    /// Field number for get and set indicating the day of the week.
    public Integer weekday;
    /// Field number for get and set indicating the count of weeks of the month.
    public Integer weekOfMonth;
    /// Field number for get and set indicating the weeks of the year.
    public Integer weekOfYear;

    private String weekdayName;

    @Override
    @SuppressWarnings("unchecked")
    public NotificationCalendarModel fromMap(Map<String, Object> arguments) {
        super.fromMap(arguments);

        era = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_ERA, Integer.class);
        year = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_YEAR, Integer.class);
        month = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_MONTH, Integer.class);
        day = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_DAY, Integer.class);
        hour = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_HOUR, Integer.class);
        minute = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_MINUTE, Integer.class);
        second = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_SECOND, Integer.class);
        millisecond = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_MILLISECOND, Integer.class);
        weekday = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_WEEKDAY, Integer.class);
        weekOfMonth = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_WEEKOFMONTH, Integer.class);
        weekOfYear = getValueOrDefault(arguments, Definitions.NOTIFICATION_SCHEDULE_WEEKOFYEAR, Integer.class);

        weekday = weekDayISO8601ToStandard(weekday);

        return this;
    }

    @Override
    public Map<String, Object> toMap() {
        Map<String, Object> returnedObject = super.toMap();

        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_ERA, era);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_YEAR, year);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_MONTH, month);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_DAY, day);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_HOUR, hour);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_MINUTE, minute);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_SECOND, second);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_MILLISECOND, millisecond);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_WEEKOFMONTH, weekOfMonth);
        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_WEEKOFYEAR, weekOfYear);

        returnedObject.put(Definitions.NOTIFICATION_SCHEDULE_WEEKDAY, weekDayStandardToISO8601(weekday));

        return returnedObject;
    }

    static Integer weekDayISO8601ToStandard(@Nullable Integer weekdayISOValue) {
        if (weekdayISOValue == null) return null;
        if (weekdayISOValue <= 0) return weekdayISOValue;
        if (weekdayISOValue > 7) return weekdayISOValue;

        if (weekdayISOValue == 7) return 1;
        return weekdayISOValue + 1;
    }

    static Integer weekDayStandardToISO8601(@Nullable Integer weekdayValue) {
        if (weekdayValue == null) return null;
        if (weekdayValue <= 0) return weekdayValue;
        if (weekdayValue > 7) return weekdayValue;

        if (weekdayValue == 1) return 7;
        return weekdayValue - 1;
    }

    @Override
    public String toJson() {
        return templateToJson();
    }

    @Override
    public NotificationCalendarModel fromJson(String json) {
        return (NotificationCalendarModel) super.templateFromJson(json);
    }

    @Override
    public void validate(Context context) throws AwesomeNotificationsException {

        if (
            era == null &&
            year == null &&
            month == null &&
            day == null &&
            hour == null &&
            minute == null &&
            second == null &&
            millisecond == null &&
            weekday == null &&
            weekOfMonth == null &&
            weekOfYear == null
        )
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "At least one time condition is required",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationCalendar");

        if (
           era != null && IntegerUtils.isBetween(era, 0, Integer.MAX_VALUE) ||
           year != null && IntegerUtils.isBetween(year, 0, Integer.MAX_VALUE) ||
           month != null && IntegerUtils.isBetween(month, 1, 12) ||
           day != null && IntegerUtils.isBetween(day, 1, 31) ||
           hour != null && IntegerUtils.isBetween(hour, 0, 23) ||
           minute != null && IntegerUtils.isBetween(minute, 0, 59) ||
           second != null && IntegerUtils.isBetween(second, 0, 59) ||
           millisecond != null && IntegerUtils.isBetween(millisecond, 0, 999) ||
           weekday != null && IntegerUtils.isBetween(weekday, 1, 7) ||
           weekOfMonth != null && IntegerUtils.isBetween(weekOfMonth, 1, 6) ||
           weekOfYear != null && IntegerUtils.isBetween(weekOfYear, 1, 53)
        )
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "The time conditions are invalid",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationCalendar");
    }

    @Override
    @Nullable
    public Calendar getNextValidDate(
            @NonNull Calendar fixedNowDate
    ) throws AwesomeNotificationsException {

        if (timeZone == null)
            throw ExceptionFactory
                    .getInstance()
                    .createNewAwesomeException(
                            TAG,
                            ExceptionCode.CODE_INVALID_ARGUMENTS,
                            "Invalid time zone",
                            ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationCalendar.timeZone");

        String cronExpression =
                (second == null ? "*" : second.toString()) + " " +
                (minute == null ? "*" : minute.toString()) + " " +
                (hour == null ? "*" : hour.toString()) + " " +
                (weekday != null ? "?" : (day == null ? "*" : day.toString())) + " " +
                (month == null ? "*" : month.toString()) + " " +
                (weekday == null ? "?" : weekday.toString()) + " " +
                (year == null ? "*" : year.toString());

        return CronUtils
                .getNextCalendar(
                        fixedNowDate,
                        cronExpression,
                        timeZone);
    }
}

File: ios/AwnCore/Classes/models/NotificationCalendarModel.swift
//
//  NotificationCalendarModel.swift
//  awesome_notifications
//
//  Created by Rafael Setragni on 09/03/21.
//

import Foundation

public class NotificationCalendarModel : NotificationScheduleModel {
    
    var _createdDate:RealDateTime?
    var _timeZone:TimeZone?
    
    public init(){}
    
    /// Initial reference date from schedule
    public var createdDate:RealDateTime? { get{
        return _createdDate
    } set(newValue){
        _createdDate = newValue
    }}
    
    /// Initial reference date from schedule
    public var timeZone:TimeZone? { get{
        return _timeZone
    } set(newValue){
        _timeZone = newValue
    }}
    
    /// Field number for get and set indicating the year.
    var year:Int?
    /// Field number for get and set indicating the month.
    var month:Int?
    /// Field number for get and set indicating the day number of the month (1-31).
    var day:Int?
    /// Field number for get and set indicating the hour of the day (0-23).
    var hour:Int?
    /// Field number for get and set indicating the minute within the hour (0-59).
    var minute:Int?
    /// Field number for get and set indicating the second within the minute (0-59).
    var second:Int?
    /// Field number for get and set indicating the day of the week (1-7).
    var weekday:Int?
    /// Field number for get and set indicating the count of weeks of the month (0-53).
    var weekOfMonth:Int?
    /// Field number for get and set indicating the weeks of the year.
    var weekOfYear:Int?
    /// Specify false to deliver the notification one time. Specify true to reschedule the notification request each time the notification is delivered.
    var repeats:Bool?
    
    public func fromMap(arguments: [String : Any?]?) -> AbstractModel? {
        
        self._createdDate =
            MapUtils<RealDateTime>.getRealDateOrDefault(
                reference: Definitions.NOTIFICATION_CREATED_DATE, arguments: arguments, defaultTimeZone: RealDateTime.utcTimeZone)
        
        self._timeZone =
            MapUtils<TimeZone>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_TIMEZONE, arguments: arguments)
        
        self.year = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_YEAR, arguments: arguments)
        self.month = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_MONTH, arguments: arguments)
        self.day = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_DAY, arguments: arguments)
        self.hour = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_HOUR, arguments: arguments)
        self.minute = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_MINUTE, arguments: arguments)
        self.second = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_SECOND, arguments: arguments)
        self.weekday = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_WEEKDAY, arguments: arguments)
        self.weekOfMonth = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_WEEKOFMONTH, arguments: arguments)
        self.weekOfYear = MapUtils<Int>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_WEEKOFYEAR, arguments: arguments)
        
        self.repeats = MapUtils<Bool>.getValueOrDefault(reference: Definitions.NOTIFICATION_SCHEDULE_REPEATS, arguments: arguments)
        
        if (self.year ?? 0) < 0 { self.year = nil }
        if (self.month ?? 0) < 0 { self.month = nil }
        if (self.day ?? 0) < 0 { self.day = nil }
        if (self.hour ?? 0) < 0 { self.hour = nil }
        if (self.minute ?? 0) < 0 { self.minute = nil }
        if (self.second ?? 0) < 0 { self.second = nil }
        if (self.weekday ?? 0) < 0 { self.weekday = nil }
        if (self.weekOfMonth ?? 0) < 0 { self.weekOfMonth = nil }
        if (self.weekOfYear ?? 0) < 0 { self.weekOfYear = nil }
        
        // https://github.com/rafaelsetragni/awesome_notifications/issues/153#issuecomment-830732722
        if(self.weekday != nil){
            self.weekday = self.weekday == 7 ? 1 : (self.weekday! + 1)
        }

        return self
    }

    public func toMap() -> [String : Any?] {
        var mapData:[String: Any?] = [:]
        
        if(_timeZone != nil)   {mapData[Definitions.NOTIFICATION_SCHEDULE_TIMEZONE] = TimeZoneUtils.shared.timeZoneToString(timeZone: self._timeZone)}
        if(_createdDate != nil){mapData[Definitions.NOTIFICATION_CREATED_DATE] = self._createdDate!.description}

        if(year != nil)        {mapData[Definitions.NOTIFICATION_SCHEDULE_YEAR]   = self.year}
        if(month != nil)       {mapData[Definitions.NOTIFICATION_SCHEDULE_MONTH]  = self.month}
        if(day != nil)         {mapData[Definitions.NOTIFICATION_SCHEDULE_DAY]    = self.day}
        if(hour != nil)        {mapData[Definitions.NOTIFICATION_SCHEDULE_HOUR]   = self.hour}
        if(minute != nil)      {mapData[Definitions.NOTIFICATION_SCHEDULE_MINUTE] = self.minute}
        if(second != nil)      {mapData[Definitions.NOTIFICATION_SCHEDULE_SECOND] = self.second}
        if(weekday != nil)     {mapData[Definitions.NOTIFICATION_SCHEDULE_WEEKDAY]     = self.weekday == 1 ? 7 : (self.weekday! - 1)}
        if(weekOfMonth != nil) {mapData[Definitions.NOTIFICATION_SCHEDULE_WEEKOFMONTH] = self.weekOfMonth}
        if(weekOfYear != nil)  {mapData[Definitions.NOTIFICATION_SCHEDULE_WEEKOFYEAR]  = self.weekOfYear}
        if(repeats != nil)     {mapData[Definitions.NOTIFICATION_SCHEDULE_REPEATS] = self.repeats}

        return mapData
    }

    public func validate() throws {
        if(
            year == nil &&
            month == nil &&
            day == nil &&
            hour == nil &&
            minute == nil &&
            second == nil &&
            weekday == nil &&
            weekOfMonth == nil &&
            weekOfYear == nil
        ){
            throw ExceptionFactory
                .shared
                .createNewAwesomeException(
                    className: NotificationIntervalModel.TAG,
                    code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                    message: "At least one time condition is required",
                    detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationCalendar")
        }

        if(!(
            IntUtils.isBetween(self.year ?? 0, min: 0, max: 99999) &&
            IntUtils.isBetween(self.month ?? 1, min: 1, max: 12) &&
            IntUtils.isBetween(self.day ?? 1, min: 1, max: 31) &&
            IntUtils.isBetween(self.hour ?? 0, min: 0, max: 23) &&
            IntUtils.isBetween(self.minute ?? 0, min: 0, max: 59) &&
            IntUtils.isBetween(self.second ?? 0, min: 0, max: 59) &&
            IntUtils.isBetween(self.weekday ?? 1, min: 1, max: 7) &&
            IntUtils.isBetween(self.weekOfMonth ?? 1, min: 1, max: 6) &&
            IntUtils.isBetween(self.weekOfYear ?? 1, min: 1, max: 53)
        )){
            throw ExceptionFactory
                .shared
                .createNewAwesomeException(
                    className: NotificationIntervalModel.TAG,
                    code: ExceptionCode.CODE_INVALID_ARGUMENTS,
                    message: "The time conditions are invalid",
                    detailedCode: ExceptionCode.DETAILED_INVALID_ARGUMENTS+".notificationCalendar")
        }
    }

    public func toDateComponents() -> DateComponents {
        let dateComponents:DateComponents = DateComponents(
            timeZone: timeZone ?? TimeZone.current,
            year: year,
            month: month,
            day: day,
            hour: hour,
            minute: minute,
            second: second,
            weekday: weekday,
            weekOfMonth: weekOfMonth,
            weekOfYear: weekOfYear
        )
        return dateComponents
    }
    
    public func getNextValidDate() -> RealDateTime? {
        let timeZone:TimeZone = self.timeZone ?? TimeZone.current
        
        let referenceDate =
            (self.repeats ?? true) ?
                RealDateTime(fromTimeZone: timeZone):
                createdDate ?? RealDateTime(fromTimeZone: timeZone)
        
        guard let nextValidDate =
                    DateUtils
                        .shared
                        .getNextValidDate(
                            fromScheduleModel: self,
                            withReferenceDate: referenceDate)
        else {
            return nil
        }
        
        return RealDateTime(
            fromDate: nextValidDate,
            inTimeZone: timeZone)
    }
    
    public func hasNextValidDate() -> Bool {
        
        let timeZone:TimeZone = self.timeZone ?? TimeZone.current
        let nowDate:RealDateTime? = RealDateTime(fromTimeZone: timeZone)
        
        let nextValidDate:RealDateTime? = getNextValidDate()
        
        return
            nil != nextValidDate &&
            nil != nowDate &&
            nextValidDate! > nowDate!
    }

    public func getUNNotificationTrigger() -> UNNotificationTrigger? {

        do {
            try validate();

            let trigger = UNCalendarNotificationTrigger(
                dateMatching: toDateComponents(),
                repeats: repeats!
            )
            
            return trigger
            
        } catch {
            Logger.e("NotificationCallendarModel", error.localizedDescription)
        }
        return nil
    }
}

